src/main/utility/tagged_ptr.c

```diff
@@ -0,0 +1,30 @@
+/*
+ * The Shadow Simulator
+ * Copyright (c) 2010-2011, Rob Jansen
+ * See LICENSE for licensing information
+ */
+
+#include "main/utility/tagged_ptr.h"
+#include "main/utility/utility.h"
+
+// two low-order bits
+const uintptr_t TAG_MASK = (1 << 2) - 1;
```

Comment:
IIRC on x86_64, objects are 8-byte-aligned, so we could use one more bit. We shouldn't end up needing it for CompatSocket, but maybe worth doing now since this is a standalone utility module.

Definitely double-check first though, or if you don't want to spend time doing that just leave a TODO comment to consider verifying that and adding one more bit. 

I see you have a `utility_assert` that should detect if I'm wrong, but I could imagine some objects being 8-byte-aligned in debug builds, but "packed" into 4-byte-alignment in release builds (where the assert is disabled).

Reply:
Good point! I've added explicit error calls here. I've also added an extra bit to the tag. I originally didn't just because we didn't need it, but there should be no harm in making it 3 bits instead of 2.